src/sentry/ddm/api.py

```diff
@@ -0,0 +1,386 @@
+            series[metric_name] = _generate_full_series(
+                start_seconds, num_intervals, interval, metric_values[0]
+            )
+            totals[metric_name] = metric_values[1]
+
+        inner_group = {
+            "by": {name: value for name, value in group_key},
+            "series": series,
+            "totals": totals,
+        }
+
+        translated_groups.append(inner_group)
+
+    translated_meta = [
+        {"name": meta_name, "type": meta_type} for meta_name, meta_type in intermediate_meta.items()
+    ]
+
+    return {
+        "intervals": intervals,
+        "groups": translated_groups,
+        "meta": translated_meta,
+        "start": start,
+        "end": end,
+    }
+
+
+def _execute_series_and_totals_query(
+    organization: Organization, use_case_id: UseCaseID, interval: int, base_query: MetricsQuery
+) -> Mapping[str, Any]:
+    dataset = Dataset.Metrics if use_case_id == UseCaseID.SESSIONS else Dataset.PerformanceMetrics
+    request = Request(
+        dataset=dataset.value,
+        query=base_query,
+        app_id="default",
+        tenant_ids={"referrer": "metrics.data", "organization_id": organization.id},
+    )
+
+    base_query.rollup = Rollup(interval=interval, granularity=_get_granularity(interval))
+    series_result = run_query(request=request)
+
+    # This is a hack, to make sure that we choose the right granularity for the totals query.
+    # This is done since for example if we query 24 hours with 1 hour interval:
+    # * For series the granularity is 1 hour since we want to aggregate on 1 hour
+    # * For totals it doesn't make sense to choose 1 hour, and instead it's better to choose 24 hours
+    series_start_seconds = series_result["start"].timestamp()
+    series_end_seconds = series_result["end"].timestamp()
+    totals_interval_seconds = int(series_end_seconds - series_start_seconds)
+
+    base_query.rollup = Rollup(totals=True, granularity=_get_granularity(totals_interval_seconds))
+    request.query = base_query
+    totals_result = run_query(request=request)
+
+    return {**series_result, "totals": totals_result["data"]}
+
+
+def run_metrics_query(
+    fields: Sequence[str],
+    query: Optional[str],
+    group_bys: Optional[Sequence[str]],
+    interval: int,
+    start: datetime,
+    end: datetime,
+    use_case_id: UseCaseID,
+    organization: Organization,
+    projects: Sequence[Project],
+):
+    # Build the basic query that contains the metadata.
+    base_query = MetricsQuery(
+        start=start,
+        end=end,
+        scope=MetricsScope(
+            org_ids=[organization.id],
+            project_ids=[project.id for project in projects],
+            use_case_id=use_case_id.value,
+        ),
+    )
+
+    # Parse all the strings and convert them to an intermediate format.
+    parsed_fields = _parse_fields(fields)
+    parsed_filters = _parse_query(query)
+    parsed_group_bys = _parse_group_by(group_bys)
```

Comment:
Please parse directly into the metrics query types, which should allow us to shorten this similar to:

```py
    base_query = MetricsQuery(
        filters=_parse_filters(query) if query else None,
        groupby=[Column(g) for g in group_bys] if group_bys else None,
        start=start,
        end=end,
        scope=...,
    )

    for timeseries in _parse_fields(fields):
        result = run_query(base_query.set_query(timeseries))
        ...
```